feat: always use snaps to resolve domains; include preinstalled ENS resolver snap

mirceanis opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

The embedded ENS resolution logic is being replaced by snaps that use `endowment:name-lookup`.
This change-set also includes a preinstalled snap that does ENS resolution including ENS multi-coin functionality.

The ENS resolver snap will use the Ethereum provider from the extension on Ethereum mainnet and ENS supported testnets and will revert to an Infura JSON-RPC when the extension is connected to other networks. On other networks it tries to resolve the [network-specific ENS address](https://eips.ethereum.org/EIPS/eip-2304) and if none exists will try to resolve the mainnet address (with a ⚠️ warning) as long as that address appears to be an externally owned account.

Most of the relevant changes are in `domains.js`.
Since the preinstalled snap uses ethers v6 while the extension depends on ethers v5.7 there are some changes in the lavamoat policies. Also, the more recent ethers version requires a different set of mocks, requiring a small change in the e2e test.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/26242?quickstart=1)

## **Related issues**

fixes https://github.com/MetaMask/MetaMask-planning/issues/2403
closes #18035
fixes #18648
fixes #22797
fixes #8556
fixes https://github.com/MetaMask/specifications/issues/11

## **Manual testing steps**

1. While on Ethereum mainnet, start a send flow
2. Type an ENS name in the recipient field(example _vitalik.eth_), pick the resulting address. (this should be identical to previous behavior)
3. Switch to a L2
4. Type an EOA ENS name in the recipient field (example _vitalik.eth_), observe an address being resolved with a ⚠️warning.
5. Type an ENS name that would resolve to a contract on mainnet (example _uniswap.eth_). Observe `No resolution for domain provided.` error.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->


https://github.com/user-attachments/assets/6e8866ef-f478-4647-a0f0-f1c13342edd1



## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

mirceanis committed:
feat: always use snaps to resolve domains; include preinstalled ENS resolver snap

fixes https://github.com/MetaMask/MetaMask-planning/issues/2403

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
chore(deps): add ethers@5.7.0 dev dependency, expected by @account-abstraction/contracts

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented:
@SocketSecurity ignore npm/@metamask/ens-resolver-snap@0.1.1

mirceanis committed:
chore(ci): manually update lavamoat policy to match the one getting generated on CI

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: guard the use of ENS snap by BUILD_FLASK flag

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented on package.json:
```diff
@@ -560,6 +561,7 @@
     "eslint-plugin-react-hooks": "^4.2.0",
     "eslint-plugin-storybook": "^0.6.15",
     "eta": "^3.2.0",
+    "ethers": "5.7.0",
```
Not directly used, but expected by `@account-abstraction/contracts` during tests.
Without this, tests would fail as the transitive ethers v6+ would get used.

mirceanis commented on ui/ducks/domains.js:
```diff
@@ -167,18 +172,20 @@ export function initializeDomainSlice() {
-      });
-    } else {
-      web3Provider = null;
+    if (!useSnapForENS) {
```
This block will disappear once https://github.com/MetaMask/snaps/issues/2621 gets fixed and `endowment:name-lookup` is no longer experimental

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

# Conflicts:
#	lavamoat/browserify/beta/policy.json
#	lavamoat/browserify/flask/policy.json
#	lavamoat/browserify/main/policy.json
#	lavamoat/browserify/mmi/policy.json

mirceanis committed:
chore: rebuild lavamoat policies

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

jonybur commented on app/scripts/metamask-controller.actions.test.js:
```diff
@@ -70,7 +70,7 @@ describe('MetaMaskController', function () {
 
   beforeAll(async function () {
-    await ganacheServer.start();
+    await ganacheServer.start({ port: 32545 });
```
Any reason why we use `32545`?

jonybur commented on ui/pages/confirmations/send/send-content/add-recipient/domain-input.container.js:
```diff
@@ -17,7 +17,7 @@ function mapDispatchToProps(dispatch) {
     lookupDomainName: debounce(
       (domainName) => dispatch(lookupDomainName(domainName)),
-      150,
+      450,
```
Why did we increase the debounce here? And why 450?

mirceanis replied:
The snaps based resolution is a little slower, and I think the same value is used in another part of the code.
I'm planning a followup PR that should clean-up the state machine and I'll add constants for these.

jonybur replied:
Ok, I get a bit of vertigo with magic numbers like these. Is it possible to find ourselves upon a race condition in this scenario?

mirceanis replied:
not really. there's a piece of code that resets the state if another lookup has already been requested.
The state machine is still a bit cumbersome and will be simplified a lot in the cleanup PR.
Most of the error states are no longer reachable, and some of the existing states are already irrelevant.

NidhiKJha reviewed (dismissed):
LGTM!

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

# Conflicts:
#	lavamoat/browserify/beta/policy.json
#	lavamoat/browserify/flask/policy.json
#	lavamoat/browserify/main/policy.json
#	lavamoat/browserify/mmi/policy.json

mirceanis committed:
chore: rebuild lavamoat policies

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

# Conflicts:
#	app/scripts/snaps/preinstalled-snaps.ts

mirceanis committed:
refactor: remove unused reducers in domains.js

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: strip flask checks for ENS snap

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: fix domain lookup condition and revert debounce interval

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
chore(deps): remove unused ethereum-ens-network-map

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
test: update mocks for e2e ENS test

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented on test/e2e/tests/transaction/ens.spec.ts:
```diff
@@ -59,17 +62,45 @@ describe('ENS', function (this: Suite) {
-        json: {
-          jsonrpc: '2.0',
-          id: '1111111111111111',
+    await mockServerJsonRpc(mockServer, [
```
A sequence of mocks are required now for the updated ethersjs resolution process

mirceanis commented on ui/ducks/domains.js:
```diff
@@ -73,43 +52,9 @@ const slice = createSlice({
-        action.payload;
+      const { resolutions, domainType, domainName } = action.payload;
       state.domainType = domainType;
-      if (state.domainType === ENS) {
```
Resolving snaps don't report errors and domainType is not relevant anymore.

hmalik88 reviewed (changes requested):
There now exist dead branches of code within the extension since we've rendered `domainType` useless, I think it's best if we just remove its usage from the following files:

1. `domains.js`
2. `domain-input-resolution-cell.tsx`
3. `domain-input-resolution-cell.stories.tsx`
4. `recipient.tsx`
5. `add-contact.component.js`

Apart from this, the changes look good.

mirceanis committed:
refactor: remove unused `domainType` prop

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: remove unused constants

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: remove unused messages

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

hmalik88 commented on ui/pages/confirmations/send/send-content/add-recipient/domain-input.component.js:
We should also remove the `IS_FLASK` check on line  171. Text from `t('recipientAddressPlaceholderFlask')` should replace the text in `t('recipientAddressPlaceholder')` and then `recipientAddressPlaceholderFlask` can be deleted.

mirceanis replied:
done

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

# Conflicts:
#	app/scripts/snaps/preinstalled-snaps.ts
#	lavamoat/browserify/beta/policy.json
#	lavamoat/browserify/flask/policy.json
#	lavamoat/browserify/main/policy.json
#	lavamoat/browserify/mmi/policy.json
#	yarn.lock

mirceanis committed:
chore: rebuild lavamoat policy files

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: remove flask checks from `domain-input.component.js`

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

hmalik88 commented:
Another thing to consider is that `isValidDomainName` is no longer really valid with snap resolutions in place. I say that we remove it entirely along with its tests. We additionally need to make the following changes:

1. Remove the `!IS_FLASK && !isValidDomainName(state.recipientInput)` check in `send.js` on line 1613.
2. Update the `validate` function in `add-contact.component.js` to the following:

```js
  validate = (input) => {
    if (isBurnAddress(input)) {
      this.setState({ error: INVALID_RECIPIENT_ADDRESS_ERROR });
    } else {
      this.setState({ error: null });
    }
  };
```

mirceanis committed:
refactor: remove unused message

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented:
> Another thing to consider is that `isValidDomainName` is no longer really valid with snap resolutions in place. I say that we remove it entirely along with its tests. We additionally need to make the following changes:

I agree with removing the test for flask there, but I don't really know how `isValidDomainName` would no longer be needed with flask resolution. Can you share an example?

mirceanis committed:
refactor: remove flask checks in send.js

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

hmalik88 commented:
> I agree with removing the test for flask there, but I don't really know how isValidDomainName would no longer be needed with flask resolution. Can you share an example?

We no longer are fencing domain resolution within flask. With snap provided domain resolution, a snap can technically resolve on any input, this includes non-tld domains e.g. `farcaster:hassan`. Keeping the `isValidDomainName` check in `send.js` prevents non-tld domains. In fact, the if condition in `send.js` should just be `isBurnAddress(state.recipientInput)` since we can't guard against `isValidHexAddress` anymore due to the possibility for the domain to be anything.

mirceanis committed:
test: update test to reflect new error state

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented:
> We no longer are fencing domain resolution within flask. With snap provided domain resolution, a snap can technically resolve on any input, this includes non-tld domains e.g. `farcaster:hassan`. Keeping the `isValidDomainName` check in `send.js` prevents non-tld domains. In fact, the if condition in `send.js` should just be `isBurnAddress(state.recipientInput)` since we can't guard against `isValidHexAddress` anymore due to the possibility for the domain to be anything.

I understand. Still, I think this is a risky path to take as then there's no real feedback given to the user if their input is a wrong checksum address, for example.
I propose that we give a little more thought to this validation and adjust it in another PR.

hmalik88 commented:
> I understand. Still, I think this is a risky path to take as then there's no real feedback given to the user if their input is a wrong checksum address, for example.
I propose that we give a little more thought to this validation and adjust it in another PR.

Looking into the `isValidHexAddress` function, it doesn't throw on strings that are not hex, I think it's fine to leave the condition as `isBurnAddress(state.recipientInput) || !isValidHexAddress...`, so at least your concern about the address is solved. I still think we should remove both the `!IS_FLASK && !isValidDomainName(state.recipientInput)` check as that still prevents snaps from providing resolution on non-tld domain input.

mirceanis committed:
fix: lookup name for any input that is not an address & validate possible addresses

snaps now handle the name resolution and there is no limit to what string they can resolve, however we are still maintaining the checksum check for what looks like an ethereum address

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented:
> I still think we should remove both the `!IS_FLASK && !isValidDomainName(state.recipientInput)` check as that still prevents snaps from providing resolution on non-tld domain input.

I updated `domain-input.component.js` and `send.js` to accept non-tld domain inputs as well.
Validation still happens if the input looks like an ethereum address

mirceanis committed:
test: adapt e2e test for non-tld domain input

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
chore(deps): bump ens-resolver-snap to v0.1.2

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis commented:
@SocketSecurity ignore npm/@metamask/ens-resolver-snap

mirceanis commented:
@SocketSecurity ignore npm/@metamask/ens-resolver-snap@0.1.2

hmalik88 commented on app/_locales/en/messages.json:
```diff
@@ -4229,9 +4229,6 @@
   "recipientAddressPlaceholder": {
     "message": "Enter public address (0x) or ENS name"
```
```suggestion
    "message": "Enter public address (0x) or domain name"
```

hmalik88 replied:
I'd prefer we make this change now, because this text is inaccurate and would confuse users.

mirceanis replied:
ok, I made the changes to the english text

mirceanis committed:
refactor: update [english] placeholder message for send/to input box

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
refactor: rename message key for input placeholder to trigger re-translation

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

# Conflicts:
#	app/_locales/de/messages.json
#	app/_locales/el/messages.json
#	app/_locales/en/messages.json
#	app/_locales/es/messages.json
#	app/_locales/fr/messages.json
#	app/_locales/hi/messages.json
#	app/_locales/id/messages.json
#	app/_locales/ja/messages.json
#	app/_locales/ko/messages.json
#	app/_locales/pt/messages.json
#	app/_locales/ru/messages.json
#	app/_locales/tl/messages.json
#	app/_locales/tr/messages.json
#	app/_locales/vi/messages.json
#	app/_locales/zh_CN/messages.json

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

# Conflicts:
#	package.json
#	yarn.lock

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

# Conflicts:
#	app/_locales/en/messages.json

mirceanis committed:
style: fix linter error

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
test: temporarily increase webdriver default timeout

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

mirceanis committed:
test: revert webdriver default timeout

This reverts commit 606b389e93e34e9602080105780565cf00c98918.

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

danjm commented:
@metamaskbot update-policies

mirceanis committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feat/2403-resolve-on-l2

# Conflicts:
#	lavamoat/browserify/beta/policy.json
#	lavamoat/browserify/flask/policy.json
#	lavamoat/browserify/main/policy.json
#	lavamoat/browserify/mmi/policy.json

mirceanis committed:
chore: rebuild policies

Signed-off-by: Mircea Nistor <mirceanis@gmail.com>

hmalik88 reviewed (approved):
LGTM

jpuri reviewed (approved):
Change in files owned by confirmations team look good.

weizman reviewed (approved):
lgtm (strictly policies):

* new globals:
  * `fetch` and `WebSocket` introduced by a different version of the pkg (so they're not new)
  * `define` is also already used previously
  * could not find other new globals that may be risky
* new `packages`
  * packages consumption seems fine too, just a reorg
* I only checked `main/policy.json`, assuming the rest are similar (worth a check @mirceanis)
